lib/middleware/gvl_instrumentation.rb

```diff
@@ -0,0 +1,31 @@
+require "gvl_timing"
+
+class GVLInstrumentation
+  def initialize(app)
+    @app = app
+  end
+
+  def call(env)
+    response = nil
+
+    before_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    before_gc_time = GC.total_time
+    timer = GVLTiming.measure do
+      response = @app.call(env)
+    end
+    total_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - before_time
+    gc_time = GC.total_time - before_gc_time
+
+    gc_ms = (gc_time / 1_000_000.0).round(2)
+    run_ms = (timer.cpu_duration * 1_000.0).round(2)
+    idle_ms = (timer.idle_duration * 1_000.0).round(2)
+    stall_ms = (timer.stalled_duration * 1_000.0).round(2)
+    io_percent = (timer.idle_duration / total_time * 100.0).round(1)
+    Rails.logger.info(
+      "[GVL] io_percent=#{io_percent}, run_ms=#{run_ms}, idle_ms=#{idle_ms}, stall_ms=#{stall_ms}, " \
+      "gc_ms=#{gc_ms}, path=#{env["PATH_INFO"]}"
```

Comment:
Not sure if `PATH_INFO` is the best, as it will have a fairly high cardinality. Perhaps `controller/action` would be better?

Reply:
You probably want controller/action. Spiders regularly map all of the site and we have a few hundred thousand URLs. You’d gather hit frequency, but basically everything is log normal on recency like you’d assume.

Reply:
> Maybe you want http verb too?

Good call. Even though it can be inferred from the action, it's easier that way.

> but I could check logs if it matters to you.

Nah, it's fine. I'm interested in requests that are actually processed by Ruby.